tensornetwork/contractors/opt_einsum_paths/path_contractors.py

```diff
@@ -1,50 +1,89 @@
 
 import functools
 import opt_einsum
-from typing import Any, Callable, Dict, Optional, List, Set
+from typing import Any, Callable, Dict, Optional, List, Set, Sequence
 from tensornetwork import network
+from tensornetwork import network_components
 from tensornetwork.contractors.opt_einsum_paths import utils
 
 
 def base(net: network.TensorNetwork,
-         algorithm: Callable[[List[Set[int]], Set[int], Dict[int, int]],
-                             List]) -> network.TensorNetwork:
-  """Base method for all `opt_einsum` contractors.
+         algorithm: Callable[[List[Set[int]], Set[int], Dict[int, int]], List],
+         output_edge_order: Sequence[network_components.Edge] = None
+        ) -> network.TensorNetwork:
+    """Base method for all `opt_einsum` contractors.
 
   Args:
     net: a TensorNetwork object. Should be connected.
     algorithm: `opt_einsum` contraction method to use.
-
+    output_edge_order: An optional list of edges. Edges of the 
+                       final node in `nodes_set` 
+                       are reordered into `output_edge_order`; 
+                       if final node has more than one edge, 
+                       `output_edge_order` must be provided.
   Returns:
     The network after full contraction.
   """
-  net.check_connected()
-  # First contract all trace edges
-  edges = net.get_all_nondangling()
-  for edge in edges:
-    if edge in net and edge.is_trace():
-      net.contract_parallel(edge)
-  if not net.get_all_nondangling():
-    # There's nothing to contract.
-    return net
 
-  # Then apply `opt_einsum`'s algorithm
-  nodes = sorted(net.nodes_set)
-  input_sets = utils.get_input_sets(net)
-  output_set = utils.get_output_set(net)
-  size_dict = utils.get_size_dict(net)
-  path = algorithm(input_sets, output_set, size_dict)
-  for a, b in path:
-    new_node = nodes[a] @ nodes[b]
-    nodes.append(new_node)
-    nodes = utils.multi_remove(nodes, [a, b])
-  return net
+    net.check_connected()
+    # First contract all trace edges
+    edges = net.get_all_nondangling()
+    for edge in edges:
+        if edge in net and edge.is_trace():
+            net.contract_parallel(edge)
+    if not net.get_all_nondangling():
+        # There's nothing to contract.
+        return net
+
+    # Then apply `opt_einsum`'s algorithm
+    nodes = sorted(net.nodes_set)
+    input_sets = utils.get_input_sets(net)
+    output_set = utils.get_output_set(net)
+    size_dict = utils.get_size_dict(net)
+    path = algorithm(input_sets, output_set, size_dict)
+    for a, b in path:
+        new_node = nodes[a] @ nodes[b]
+        nodes.append(new_node)
+        nodes = utils.multi_remove(nodes, [a, b])
+
+    # if the final node has more than one edge,
+    # output_edge_order has to be specified
+    final_node = net.get_final_node()
+    if (len(final_node.edges) <= 1) and (output_edge_order is None):
+        output_edge_order = list(
+            (net.get_all_edges() - net.get_all_nondangling()))
```

Comment:
Oh man we really should just add `net.get_all_dangling()`...

Reply:
Yes!